boot/freeldr/freeldr/lib/fs/btrfs.c

```diff
@@ -0,0 +1,1210 @@
+            rd = btrfs_read_extent_reg(&path, extent, offset_in_extent, size, buf);
+        }
+
+        if (rd == -1ULL) {
+            ERR("Error while reading extent\n");
+            seek_pointer = -1ULL;
+            goto out;
+        }
+
+        buf += rd;
+        seek_pointer += rd;
+        size -= rd;
+        TRACE("file_read size=%llu rd=%llu seek_pointer=%llu\n", size, rd, seek_pointer);
+
+        if (!size)
+            break;
+    } while (!(res = next_slot(&key, &path)));
+
+    if (res) {
+        seek_pointer = -1ULL;
+        goto out;
+    }
+
+    seek_pointer -= offset;
+    out:
+    free_path(&path);
+    return seek_pointer;
+}
+
+
+ARC_STATUS BtrFsClose(ULONG FileId)
+{
+    pbtrfs_file_info phandle = FsGetDeviceSpecific(FileId);
+    TRACE("BtrFsClose %lu\n", FileId);
+
+    FrLdrTempFree(phandle, TAG_BTRFS_FILE);
+
+    return ESUCCESS;
+}
+
+ARC_STATUS BtrFsGetFileInformation(ULONG FileId, FILEINFORMATION* Information)
+{
+    pbtrfs_file_info phandle = FsGetDeviceSpecific(FileId);
+
+    TRACE("BtrFsGetFileInformation %lu\n", FileId);
+
+    RtlZeroMemory(Information, sizeof(FILEINFORMATION));
+    Information->EndingAddress.QuadPart = phandle->inode.size;
+    Information->CurrentAddress.QuadPart = phandle->position;
+
+    TRACE("BtrFsGetFileInformation() FileSize = %llu\n",
+          Information->EndingAddress.QuadPart);
+    TRACE("BtrFsGetFileInformation() FilePointer = %llu\n",
+          Information->CurrentAddress.QuadPart);
+
+    return ESUCCESS;
+}
+
+ARC_STATUS BtrFsOpen(CHAR* Path, OPENMODE OpenMode, ULONG* FileId)
+{
+    u64 inr;
+    u8 type;
+
+    btrfs_file_info temp_file_info;
+    pbtrfs_file_info phandle;
+
+    TRACE("BtrFsOpen %s\n", Path);
+
+    if (OpenMode != OpenReadOnly)
+        return EACCES;
+
+    inr = btrfs_lookup_path(&BtrFsInfo->FsRoot, BtrFsInfo->FsRoot.root_dirid, Path, &type, &temp_file_info.inode, 40);
+
+    if (inr == -1ULL) {
+        TRACE("Cannot lookup file %s\n", Path);
+        return ENOENT;
+    }
+
+    if (type != BTRFS_FT_REG_FILE) {
+        TRACE("Not a regular file: %s\n", Path);
+        return EISDIR;
+    }
+
+    TRACE("found inode inr=%llu size=%llu\n", inr, temp_file_info.inode.size);
+
+    temp_file_info.inr = inr;
+    temp_file_info.position = 0;
+
+    phandle = FrLdrTempAlloc(sizeof(btrfs_file_info), TAG_BTRFS_FILE);
+    if (!phandle)
+        return ENOMEM;
+
+    memcpy(phandle, &temp_file_info, sizeof(btrfs_file_info));
+
+    FsSetDeviceSpecific(*FileId, phandle);
+    return ESUCCESS;
+}
+
+ARC_STATUS BtrFsRead(ULONG FileId, VOID* Buffer, ULONG N, ULONG* Count)
+{
+    pbtrfs_file_info phandle = FsGetDeviceSpecific(FileId);
+    u64 rd;
+
+    TRACE("BtrFsRead %lu, size=%lu \n", FileId, N);
+
+    if (!N)
+        N = phandle->inode.size;
+
+    if (N > phandle->inode.size)
+        N = phandle->inode.size;
+
+    rd = btrfs_file_read(&BtrFsInfo->FsRoot, phandle->inr, phandle->position, N, Buffer);
+    if (rd == -1ULL) {
```

Comment:
Please use the non-K&R style : https://www.reactos.org/wiki/Coding_Style#Braces
(This comment applies to all of this file, and possibly to the btrfs.h file as well, unless in the btrfs.h the structs come directly from 3rd-party, in which case they may be left untouched).

Reply:
It has been talked about more than once. It would be possible, but there's more than one style in the code depending on the module and where it comes from (win32ss style is different from NT kernel style, and 3rdparty libs have their own, and wine has their own...), so that would mean two things:
1. we'd have to include style settings per file (in some comment line in the header or such, and
2. we'd have to make the style inside a file consistent, which is not always the case.

1 is annoying because it introduces a new line into every file, while 2 is annoying because it introduces a lot of whitespace changes to existing code, making it more annoying to follow the file's history.